### PR TITLE
feat(cli): add new update command to cli

### PR DESCRIPTION
**Pull request by yeshamavani**
gh-528

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## Checklist:

- [x] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/cli/src/generators/update/index.ts [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+      return false;
+    }
+  }
+
+  private async _checkDependencies() {
+    const packageJson = this.fs.readJSON(
+      this.destinationPath(packageJsonFile),
+    ) as AnyObject;
+
+    // const pkgDeps: {
```
Comment (akshatdubeysf) [inline packages/cli/src/generators/update/index.ts:110]: ???

File: packages/cli/src/generators/update/index.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+      }
+      if (answer && answer.decision === 'upgrade') {
+        await this._updateDependencies();
+        return true;
+      }
+    }
+  }
+
+  private _compare(tempDependency: string, dependency: string) {
+    if (dependency !== tempDependency) {
```
Comment (akshatdubeysf) [inline packages/cli/src/generators/update/index.ts:98]: return dependency !== tempDependency

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: packages/cli/src/generators/update/index.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+const packageJsonFile = 'package.json';
+
+export default class UpdateGenerator extends BaseUpdateGenerator<UpdateOptions> {
+  constructor(public args: string[], public opts: UpdateOptions) {
+    super(args, opts);
+  }
+
+  initializing() {
+    if (!this.fs.exists(join(this.destinationPath(), 'lerna.json'))) {
```
Comment (samarpan-b) [inline packages/cli/src/generators/update/index.ts:19]: why is this needed at rot only ?

File: packages/cli/src/generators/update/index.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+    for (const type of types) {
+      this.destinationRoot(monoRepo);
+      const folders = this._getDirectories(
+        this.destinationRoot(join('.', type)),
+      );
+      for (const folder of folders) {
+        this.destinationRoot(join(monoRepo, type, folder));
+
+        await super.checkLoopBackProject();
```
Comment (samarpan-b) [inline packages/cli/src/generators/update/index.ts:36]: await inside loop is simply bad programming

File: packages/cli/src/generators/update/index.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  async updateAllProjects() {
+    const types = ['facades', 'services', 'packages'];
+    const monoRepo = this.destinationPath();
+
+    for (const type of types) {
+      this.destinationRoot(monoRepo);
+      const folders = this._getDirectories(
+        this.destinationRoot(join('.', type)),
+      );
+      for (const folder of folders) {
```
Comment (samarpan-b) [inline packages/cli/src/generators/update/index.ts:33]: cant we reduce time complexity here ? Its O(n^2) right now. Try reducing it to O(n)

File: packages/cli/src/generators/update/index.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+      }
+    }
+
+    // lerna bootstrap at the end
+    this.destinationRoot(monoRepo);
+    this.spawnCommandSync('lerna', ['bootstrap']);
+  }
+
+  private _getDirectories(folderPath: string) {
+    return readdirSync(folderPath, {withFileTypes: true})
```
Comment (samarpan-b) [inline packages/cli/src/generators/update/index.ts:57]: shouldn't we use async method here ?

File: packages/cli/src/generators/update/index.ts [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+      const devDep = pkgDeps.devDependencies[d];
+      const peerDep = pkgDeps.peerDependencies[d];
+      if (!dep && !devDep && !peerDep) continue;
+
+      const tempDependency = tempDeps[d];
+      if (dep && this._compare(tempDependency, dep)) {
+        depsToUpdate.dependencies[d] = tempDependency;
+        found = true;
+      }
+      if (devDep && this._compare(tempDependency, devDep)) {
```
Comment (samarpan-b) [inline packages/cli/src/generators/update/index.ts:120]: this._compare is invoked multiple times in the same fn. We can invoke once and use its value. This improves performance of the code.

File: packages/cli/src/types.ts [hunk 8]
```diff
@@ -7,6 +7,18 @@ export interface ProjectInfo {
     [key: string]: string;
   };
 }
+
+// sonarignore:start
+export interface DependencyType {
```
Comment (samarpan-b) [inline packages/cli/src/types.ts:12]: move types specific to sl update  command into separate file.

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: packages/cli/src/generators/update/index.ts (hunk 4 above)
Comment (yeshamavani) [reply packages/cli/src/generators/update/index.ts:36]: we will need this as not able to achieve desired flow without it

File: packages/cli/src/generators/update/index.ts (hunk 5 above)
Comment (yeshamavani) [reply packages/cli/src/generators/update/index.ts:33]: actually the outer loop is fixed so it is like m*n = 3n,

File: packages/cli/src/generators/update/index.ts (hunk 7 above)
Comment (yeshamavani) [reply packages/cli/src/generators/update/index.ts:120]: removed the function checking directly

File: packages/cli/src/types.ts (hunk 8 above)
Comment (yeshamavani) [reply packages/cli/src/types.ts:12]: done

**Review by akshatdubeysf — APPROVED** [review]

**Pull request merged**